core/src/main/java/org/apache/calcite/util/format/postgresql/PostgresqlDateTimeFormatter.java

```diff
@@ -308,4 +481,362 @@ public static String toChar(String formatString, ZonedDateTime dateTime) {
+        .truncatedTo(ChronoUnit.DAYS);
+
+    DateCalendarEnum calendar = DateCalendarEnum.NONE;
+    boolean containsCentury = false;
+    for (ChronoUnitEnum unit : dateParts.keySet()) {
+      if (unit.getCalendars().size() == 1) {
+        DateCalendarEnum unitCalendar = unit.getCalendars().iterator().next();
+        if (unitCalendar != DateCalendarEnum.NONE) {
+          calendar = unitCalendar;
+          break;
+        }
+      } else if (unit == ChronoUnitEnum.CENTURIES) {
+        containsCentury = true;
+      }
+    }
+
+    if (calendar == DateCalendarEnum.NONE && containsCentury) {
+      calendar = DateCalendarEnum.GREGORIAN;
+    }
+
+    switch (calendar) {
+    case NONE:
+      constructedDateTime = constructedDateTime
+          .withYear(1)
+          .withMonth(1)
+          .withDayOfMonth(1);
+      break;
+    case GREGORIAN:
+      constructedDateTime = updateWithGregorianFields(constructedDateTime, dateParts);
+      break;
+    case ISO_8601:
+      constructedDateTime = updateWithIso8601Fields(constructedDateTime, dateParts);
+      break;
+    case JULIAN:
+      final Long julianDays = dateParts.get(ChronoUnitEnum.DAYS_JULIAN);
+      if (julianDays != null) {
+        constructedDateTime = constructedDateTime.with(JulianFields.JULIAN_DAY, julianDays);
+      }
+      break;
+    }
+
+    constructedDateTime = updateWithTimeFields(constructedDateTime, dateParts);
+
+    if (dateParts.containsKey(ChronoUnitEnum.TIMEZONE_HOURS)
+        || dateParts.containsKey(ChronoUnitEnum.TIMEZONE_MINUTES)) {
+      final int hours = dateParts.getOrDefault(ChronoUnitEnum.TIMEZONE_HOURS, 0L)
+          .intValue();
+      final int minutes = dateParts.getOrDefault(ChronoUnitEnum.TIMEZONE_MINUTES, 0L)
+          .intValue();
+
+      return ZonedDateTime.of(constructedDateTime, ZoneOffset.ofHoursMinutes(hours, minutes));
+    }
+
+    return ZonedDateTime.of(constructedDateTime, zoneId);
+  }
+
+  private static LocalDateTime updateWithGregorianFields(LocalDateTime dateTime,
+      Map<ChronoUnitEnum, Long> dateParts) {
+    LocalDateTime updatedDateTime = dateTime.withYear(getGregorianYear(dateParts)).withDayOfYear(1);
+
+    if (dateParts.containsKey(ChronoUnitEnum.MONTHS_IN_YEAR)) {
+      updatedDateTime =
+          updatedDateTime.withMonth(dateParts.get(ChronoUnitEnum.MONTHS_IN_YEAR).intValue());
+    }
+
+    if (dateParts.containsKey(ChronoUnitEnum.DAYS_IN_MONTH)) {
+      updatedDateTime =
+          updatedDateTime.withDayOfMonth(dateParts.get(ChronoUnitEnum.DAYS_IN_MONTH).intValue());
+    }
+
+    if (dateParts.containsKey(ChronoUnitEnum.WEEKS_IN_MONTH)) {
+      updatedDateTime =
+          updatedDateTime.withDayOfMonth(
+              dateParts.get(ChronoUnitEnum.WEEKS_IN_MONTH).intValue() * 7 - 6);
+    }
+
+    if (dateParts.containsKey(ChronoUnitEnum.WEEKS_IN_YEAR)) {
+      updatedDateTime =
+          updatedDateTime.withDayOfYear(
```

Comment:
this doesn't look right either

Reply:
This is also to match PostgreSQL behaviour.

```
postgres=# select to_date('2024-15', 'YYYY-WW');
  to_date
------------
 2024-04-08
(1 row)

postgres=# select to_date('2024-99', 'YYYY-DDD');
  to_date
------------
 2024-04-08
(1 row)
```